File: 2019/python/day12.py
import itertools
import math

from utils import get_lines, read_data
from utils.iterables import cycle_detect


def parse_position(line):
    return [int(s[2:]) for s in line[1:-1].split(", ")]


def lcm(values):
    result = values[0]
    for value in values[1:]:
        result *= value // math.gcd(value, result)
    return result


def abs_sum(l):
    return sum(abs(x) for x in l)


class Moon:
    def __init__(self, position):
        self.position = position
        self.velocity = [0, 0, 0]

    def update_velocity(self, others):
        for other in others:
            for i, (a, b) in enumerate(zip(self.position, other.position)):
                self.velocity[i] += 0 if a == b else (1 if a < b else -1)

    def update_position(self):
        for i, v in enumerate(self.velocity):
            self.position[i] += v

    @property
    def energy(self):
        return abs_sum(self.position) * abs_sum(self.velocity)

    def __repr__(self):
        return f"Moon(position={self.position}, velocity={self.velocity})"


class PlanetarySystem:
    def __init__(self, moons):
        self.moons = moons

    def step(self):
        for moon in self.moons:
            moon.update_velocity(self.moons)
        for moon in self.moons:
            moon.update_position()

    def state(self):
        """Returns state as a tuple by dimension."""
        return tuple(
            tuple((moon.position[i], moon.velocity[i]) for moon in self.moons)
            for i in range(3)
        )

    @property
    def energy(self):
        return sum(moon.energy for moon in self.moons)


def p1(system):
    for i in range(1000):
        system.step()

    return system.energy


def p2(system):
    def states():
        while True:
            yield system.state()
            system.step()

    state_iterators = itertools.tee(states(), 3)
    cycles = [cycle_detect(state[i] for state in it) for (i, it) in enumerate(state_iterators)]
    return max(start for (start, end) in cycles) + lcm([end - start for (start, end) in cycles])


def run_tests():
    positions_a = ((-1, 0, 2), (2, -10, -7), (4, -8, 8), (3, 5, -1))
    positions_b = ((-8, -10, 0), (5, 5, 10), (2, -7, 3), (9, -8, -3))

    system = PlanetarySystem([Moon(list(p)) for p in positions_a])
    for i in range(10):
        system.step()
    assert system.energy == 179

    system = PlanetarySystem([Moon(list(p)) for p in positions_b])
    for i in range(100):
        system.step()
    assert system.energy == 1940

    system = PlanetarySystem([Moon(list(p)) for p in positions_a])
    assert p2(system) == 2772

    system = PlanetarySystem([Moon(list(p)) for p in positions_b])
    assert p2(system) == 4686774924


if __name__ == "__main__":
    run_tests()
    print("All tests passed")

    data = read_data(12)
    system = PlanetarySystem([Moon(parse_position(line)) for line in get_lines(data)])
    print("Part 1: {}".format(p1(system)))
    print("Part 2: {}".format(p2(system)))

File: 2019/rust/src/bin/day12.rs
extern crate num_integer;

use aoc::aoc_error::AOCError;
use aoc::iterators::cycle_detect;
use aoc::point::{Axis3D, Point3D};
use std::iter;
use std::str::FromStr;

type Result<T> = ::std::result::Result<T, Box<dyn ::std::error::Error>>;
type Point = Point3D<i64>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Moon {
    position: Point,
    velocity: Point,
}

impl Moon {
    fn new(position: Point) -> Moon {
        Moon {
            position,
            velocity: Point { x: 0, y: 0, z: 0 },
        }
    }

    fn update_velocity(&mut self, other: &Moon) {
        for axis in Axis3D::iter() {
            self.velocity[axis] += -(self.position[axis].cmp(&other.position[axis]) as i64);
        }
    }

    fn update_position(&mut self) {
        for axis in Axis3D::iter() {
            self.position[axis] += self.velocity[axis];
        }
    }

    fn energy(&self) -> i64 {
        (self.position.x.abs() + self.position.y.abs() + self.position.z.abs())
            * (self.velocity.x.abs() + self.velocity.y.abs() + self.velocity.z.abs())
    }
}

impl FromStr for Moon {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts = &s[1..s.len() - 1]
            .splitn(3, ", ")
            .map(|s| s[2..].parse())
            .collect::<std::result::Result<Vec<i64>, _>>()?;
        if parts.len() != 3 {
            return Err(AOCError::new("Failed to parse moon."))?;
        }

        Ok(Moon::new(Point {
            x: parts[0],
            y: parts[1],
            z: parts[2],
        }))
    }
}

#[derive(Debug, Clone)]
struct PlanetarySystem {
    moons: Vec<Moon>,
}

impl FromStr for PlanetarySystem {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        Ok(PlanetarySystem {
            moons: s.lines().map(&str::parse).collect::<Result<_>>()?,
        })
    }
}

impl PlanetarySystem {
    fn step(&mut self) {
        // iterate over every pair of moons and update velocity.
        for i in 0..self.moons.len() {
            let (a, b) = self.moons.split_at_mut(i);
            let moon_b = &mut b[0];
            for moon_a in a {
                moon_a.update_velocity(moon_b);
                moon_b.update_velocity(moon_a);
            }
        }
        for moon in &mut self.moons {
            moon.update_position();
        }
    }

    fn energy(&self) -> i64 {
        self.moons.iter().map(Moon::energy).sum()
    }

    fn states(&self) -> impl Iterator<Item = PlanetarySystem> {
        iter::successors(Some(self.clone()), |system| {
            let mut system = system.clone();
            system.step();
            Some(system)
        })
    }
}

fn cycle_length(system: &PlanetarySystem) -> Result<usize> {
    let cycles = Axis3D::iter()
        .filter_map(|axis| {
            cycle_detect(system.states().map(|system| {
                system
                    .moons
                    .iter()
                    .map(|m| (m.position[axis], m.velocity[axis]))
                    .collect::<Vec<_>>()
            }))
        })
        .collect::<Vec<_>>();
    if cycles.len() != 3 {
        Err(AOCError::new("Expected to find 3 cycles"))?;
    }

    let cycle_start = cycles.iter().map(|c| c.start).max().unwrap();
    let cycle_length = cycles
        .iter()
        .fold(1, |lcm, cycle| num_integer::lcm(lcm, cycle.length));

    Ok(cycle_start + cycle_length)
}

fn part1(mut system: PlanetarySystem) -> i64 {
    for _ in 0..1000 {
        system.step();
    }

    system.energy()
}

fn part2(system: &PlanetarySystem) -> Result<usize> {
    cycle_length(system)
}

fn main() -> Result<()> {
    let input = aoc::utils::get_input()?;
    let system: PlanetarySystem = input.parse()?;

    println!("Part 1: {}", part1(system.clone()));
    println!("Part 2: {}", part2(&system)?);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn energy_test_1() {
        let input = "<x=-1, y=0, z=2>\n\
                     <x=2, y=-10, z=-7>\n\
                     <x=4, y=-8, z=8>\n\
                     <x=3, y=5, z=-1>";
        let mut system: PlanetarySystem = input.parse().unwrap();
        for _ in 0..10 {
            system.step();
        }
        assert_eq!(system.energy(), 179);
    }

    #[test]
    fn energy_test_2() {
        let input = "<x=-8, y=-10, z=0>\n\
                     <x=5, y=5, z=10>\n\
                     <x=2, y=-7, z=3>\n\
                     <x=9, y=-8, z=-3>";
        let mut system: PlanetarySystem = input.parse().unwrap();
        for _ in 0..100 {
            system.step();
        }
        assert_eq!(system.energy(), 1940);
    }

    #[test]
    fn cycle_test_1() {
        let input = "<x=-1, y=0, z=2>\n\
                     <x=2, y=-10, z=-7>\n\
                     <x=4, y=-8, z=8>\n\
                     <x=3, y=5, z=-1>";
        let system: PlanetarySystem = input.parse().unwrap();
        assert_eq!(cycle_length(&system).unwrap(), 2772);
    }

    #[test]
    fn cycle_test_2() {
        let input = "<x=-8, y=-10, z=0>\n\
                     <x=5, y=5, z=10>\n\
                     <x=2, y=-7, z=3>\n\
                     <x=9, y=-8, z=-3>";
        let system: PlanetarySystem = input.parse().unwrap();
        assert_eq!(cycle_length(&system).unwrap(), 4686774924);
    }
}
